src/context.js

```diff
@@ -0,0 +1,64 @@
+'use strict'
+
+function clone(oldObject) {
+  // Basis.
+  if (!(oldObject instanceof Object)) {
+    return oldObject
+  }
+
+  var clonedObject
+
+  // Filter out special objects.
+  var Constructor = oldObject.constructor
+  switch (Constructor) {
+    // Implement other special objects here.
+  case Promise:
+    clonedObject = oldObject.then()
+    break
+  case Date:
+    clonedObject = new Constructor(oldObject.getTime())
+    break
+  default:
+    clonedObject = new Constructor()
+  }
+
+  // Clone each property.
+  for (var prop in oldObject) {
+    clonedObject[prop] = clone(oldObject[prop])
+  }
+
+  return clonedObject
+}
+
+function Context(generateLog, oldContext) {
+  let context = clone(oldContext)
```

Comment:
I'd like to see this tested against the serverless framework. We had edge-cases using a clone function before with it... I think we should be safe here, but lets make sure before we merge?